File: src/main/java/com/anod/appwatcher/market/CompositeStateEndpoint.java
package com.anod.appwatcher.market;

import com.android.volley.VolleyError;

/**
 * @author algavris
 * @date 27/08/2016.
 */

public class CompositeStateEndpoint extends CompositeEndpoint implements PlayStoreEndpoint.Listener {
    private int mActiveId = -1;
    protected final CompositeStateEndpoint.Listener mListener;

    public CompositeStateEndpoint(Listener listener) {
        mListener = listener;
    }

    @Override
    public void add(int id, PlayStoreEndpoint endpoint) {
        super.add(id, endpoint);
        if (mActiveId == -1)
        {
            mActiveId = id;
        }
        endpoint.setListener(this);
    }

    public PlayStoreEndpointBase active()
    {
        return (PlayStoreEndpointBase) get(mActiveId);
    }

    public CompositeStateEndpoint setActive(int id) {
        mActiveId = id;
        return this;
    }

    @Override
    public void setListener(PlayStoreEndpoint.Listener listener) {
        super.setListener(this);
    }

    @Override
    public void onDataChanged() {
        mListener.onDataChanged(mActiveId, active());
    }

    @Override
    public void onErrorResponse(VolleyError error) {
        mListener.onErrorResponse(mActiveId, active(), error);
    }

    @Override
    public void startAsync() {
        active().startAsync();
    }

    @Override
    public void startSync() {
        active().startSync();
    }

    public interface Listener
    {
        void onDataChanged(int id, PlayStoreEndpointBase endpoint);
        void onErrorResponse(int id, PlayStoreEndpointBase endpoint, VolleyError error);
    }
}

File: src/main/java/com/anod/appwatcher/market/CompositeStateEndpoint.kt
package com.anod.appwatcher.market

import com.android.volley.VolleyError

/**
 * @author algavris
 * *
 * @date 27/08/2016.
 */

class CompositeStateEndpoint(val compositeListener: Listener) : CompositeEndpoint(), PlayStoreEndpoint.Listener {
    private var mActiveId = -1

    override fun add(id: Int, endpoint: PlayStoreEndpoint) {
        super.add(id, endpoint)
        if (mActiveId == -1) {
            mActiveId = id
        }
        endpoint.listener = this
    }

    fun active(): PlayStoreEndpointBase {
        return get(mActiveId) as PlayStoreEndpointBase
    }

    fun setActive(id: Int): CompositeStateEndpoint {
        mActiveId = id
        return this
    }

    override fun onDataChanged() {
        compositeListener.onDataChanged(mActiveId, active())
    }

    override fun onErrorResponse(error: VolleyError) {
        compositeListener.onErrorResponse(mActiveId, active(), error)
    }

    override fun startAsync() {
        active().startAsync()
    }

    override fun startSync() {
        active().startSync()
    }

    interface Listener {
        fun onDataChanged(id: Int, endpoint: PlayStoreEndpointBase)
        fun onErrorResponse(id: Int, endpoint: PlayStoreEndpointBase, error: VolleyError)
    }
}
